### DISCUSSION TITLE
Why not go full WASM?

### DISCUSSION THREAD

**Discussion by erichulburd** [input-output-hk/cardano-js-sdk#381] [Ideas]
I see this library is built on top of Rust generated from CDDL: https://github.com/Emurgo/cardano-serialization-lib.

There are a number of Rust web frameworks entering the game: Yew is a popular one, Dioxus is one I'm using in a current project.

I am currently trying to look for particular assets in a user's connected wallet. The Typescript may look something like this:

```ts
import { CSL, Cardano, cslToCore } from '@cardano-sdk/core';

export const listAssets = async (walletName: string): Promise<string[]> => {
    const wallet = await window.cardano[walletName]?.enable(); 
    const utxos = (await wallet?.getUtxos()) || [];
    const transactions = cslToCore.
        utxo(utxos!.map(utxo => CSL.TransactionUnspentOutput.from_bytes(Buffer.from(utxo, 'hex'))));
    const txs = transactions.flatMap(x => x).filter(isTxOut);
    return Array.from(
        txs
        .map(tx => Object.keys(tx.value.assets || []))
        .flatMap(x => x)
        .reduce((set, assetId, _idx) => {
            set.add(assetId);
            return set;
        }, new Set<string>())
    );
};
```

When I want to call this from a Rust framework, this data is serialized to/ from WASM, to ultimately get passed to WASM again IIUC.

Why not provide the CSL functions directly in Rust and then support Typescript calls on top of that? In addition to _potential_ performance gains, it would save Rust WASM framework developers from having to deal with painful webpack configuration to get `@cardano-sdk/core` to compile.

**Comment by mkazlauskas** [comment]
`cardano-js-sdk` is intended for js/ts devs. If your application is written in Rust, then you might be able to use [cardano-serialization-lib crate](https://crates.io/crates/cardano-serialization-lib) directly, deserializing objects returned from cip30 using `from_bytes()` in your Rust app.

Side note: your `listAssets` function would be more efficient if it used `wallet.getBalance` instead of `getUtxos`.

**Comment by erichulburd** [comment]
@mkazlauskas Thanks - I didn't realize `wallet.getBalance` contains all native assets -  it's simplified my implementation.

I hit a couple of build issues using the Emurgo crate (ie it's really made for Typescript bindings rather than Rust WASM frameworks ), so I ended up using my own fork (ie [erichulburd/cardano-serialization-lib](https://github.com/erichulburd/cardano-serialization-lib)).

Thanks for the response.

**Reply by rhyslbw** [reply]
Please mark @mkazlauskas' response as the answer if you agree it addresses your original question :+1:
